File: libs/power-log-parser/src/lib/parsers/battlegrounds-reroll-parser.ts
import { BlockType, CardIds } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Action, FullEntity, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsRerollParser implements ActionParser {
	readonly ParserName = 'BattlegroundsRerollParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === NodeType.Action &&
			(node.Object as Action).Type === (BlockType.POWER as number) &&
			this.GameState.CurrentEntities.has((node.Object as Action).Entity) &&
			(this.GameState.CurrentEntities.get((node.Object as Action).Entity)!.CardId ===
				CardIds.Refresh_TB_BaconShop_1p_Reroll_Button ||
				this.GameState.CurrentEntities.get((node.Object as Action).Entity)!.CardId ===
					CardIds.Refresh_TB_BaconShop_8p_Reroll_Button)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const fullEntities = action.Data.filter((data) => data instanceof FullEntity);
		if (fullEntities.length === 0) {
			return null;
		}

		return [
			GameEventProvider.Create(
				(node.Object as Action).TimeStamp,
				'BATTLEGROUNDS_REROLL',
				() => ({
					Type: 'BATTLEGROUNDS_REROLL',
					Value: {},
				}),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsRerollParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;
using System;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsRerollParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsRerollParser(ParserState ParserState, StateFacade stateFacade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = stateFacade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(Action)
                && (node.Object as Action).Type == (int)BlockType.POWER
                && GameState.CurrentEntities.ContainsKey((node.Object as Action).Entity)
                && (
                    GameState.CurrentEntities[((node.Object as Action).Entity)].CardId == Refresh_TB_BaconShop_1p_Reroll_Button
                    || GameState.CurrentEntities[((node.Object as Action).Entity)].CardId == Refresh_TB_BaconShop_8p_Reroll_Button);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Action;
            var fullEntities = action.Data.Where(data => data is FullEntity).ToList();
            if (fullEntities.Count() == 0)
            {
                return null;
            }

            return new List<GameEventProvider> { GameEventProvider.Create(
                    (node.Object as Action).TimeStamp,
                     "BATTLEGROUNDS_REROLL",
                    () => new GameEvent
                    {
                        Type = "BATTLEGROUNDS_REROLL",
                        Value = new
                        {
                        }
                    },
                    true,
                    node)
                };
        }
    }
}
